openpower/patches/vesnin-patches/petitboot/petitboot-1001-discover-platform-powerpc-fix-bug-with-ipmi-boot-mai.patch

```diff
@@ -0,0 +1,280 @@
+-		if (resp_len < 3) {
+-			pb_log("platform: unexpected length (%d) in "
+-					"boot options mailbox response\n",
+-					resp_len);
+-			return -1;
+-		}
+-
+-		if (resp_len == 4) {
+-			pb_debug_fn("block %hu empty\n", block);
+-			return 0;
+-		}
++	if (resp_len > sizeof(ipmi_mbox_response_t)) {
++		pb_debug("platform: invalid mailbox response size!\n");
++		return -1;
++	}
+ 
+-		blocksize = sizeof(resp) - 3;
+-		pb_debug_fn("Mailbox block %hu returns only %zu bytes in block\n",
+-				block, blocksize);
++	if (resp_len < 3) {
++		pb_log("platform: unexpected length (%d) in "
++				"boot options mailbox response\n",
++				resp_len);
++		return -1;
+ 	}
+ 
+-	debug_buf = format_buffer(platform, resp, resp_len);
++	blocksize = resp_len - (sizeof(ipmi_mbox_response_t) - sizeof(mbox_t));
++	pb_debug_fn("Mailbox block %hu returns only %zu bytes in block\n",
++			block, blocksize);
++
++	debug_buf = format_buffer(platform, (uint8_t*)&ipmi_mbox_resp, resp_len);
+ 	pb_debug_fn("IPMI bootdev mailbox block %hu:\n%s\n", block, debug_buf);
+ 	talloc_free(debug_buf);
+ 
+-	if (resp[0] != 0) {
++	if (ipmi_mbox_resp.cc != 0) {
+ 		pb_log("platform: non-zero completion code %d from IPMI req\n",
+-				resp[0]);
++				ipmi_mbox_resp.cc);
+ 		return -1;
+ 	}
+ 
+ 	/* check for correct parameter version */
+-	if ((resp[1] & 0xf) != 0x1) {
++	if ((ipmi_mbox_resp.param_version & 0xf) != 0x1) {
+ 		pb_log("platform: unexpected version (0x%x) in "
+-				"boot mailbox response\n", resp[0]);
++				"boot mailbox response\n", ipmi_mbox_resp.param_version);
+ 		return -1;
+ 	}
+ 
+ 	/* check for valid paramters */
+-	if (resp[2] & 0x80) {
++	if (ipmi_mbox_resp.param_valid & 0x80) {
+ 		pb_debug("platform: boot mailbox parameters are invalid/locked\n");
+ 		return -1;
+ 	}
+ 
+-	memcpy(buf, &resp[3], blocksize);
++	if (block != ipmi_mbox_resp.block_selector) {
++		pb_debug("platform: returned boot mailbox block doesn't match "
++				  "requested\n");
++		return -1;
++	}
++
++	if (blocksize == 0) {
++		pb_debug_fn("block %hu empty\n", block);
++		return -1;
++	}
++
++	memcpy(mailbox, &ipmi_mbox_resp.mbox, blocksize);
+ 
+ 	return blocksize;
+ }
+@@ -511,15 +543,15 @@ static int get_ipmi_boot_mailbox(struct platform_powerpc *platform,
+ 		char **buf)
+ {
+ 	char *mailbox_buffer, *prefix;
+-	const size_t blocksize = 16;
+-	char block_buffer[blocksize];
+-	size_t mailbox_size;
++	mbox_t mailbox;
++	uint8_t *boot_opt_data;
++	size_t buffer_size;
+ 	int content_size;
+ 	uint8_t i;
+ 	int rc;
+ 
+ 	mailbox_buffer = NULL;
+-	mailbox_size = 0;
++	buffer_size = 0;
+ 
+ 	/*
+ 	 * The BMC may hold up to 255 blocks of data but more likely the number
+@@ -527,8 +559,8 @@ static int get_ipmi_boot_mailbox(struct platform_powerpc *platform,
+ 	 * on higher numbers.
+ 	 */
+ 	for (i = 0; i < UCHAR_MAX; i++) {
+-		rc = get_ipmi_boot_mailbox_block(platform, block_buffer, i);
+-		if (rc < 3 && i == 0) {
++		rc = get_ipmi_boot_mailbox_block(platform, &mailbox, i);
```

Comment:
I would rename this `rc` to something more relevant like `block_size`.
Also it seems like the scope of this `rc`/`block_size` variable can be decreased to just the `for` block.

Reply:
Done